### PR TITLE
Job patterns documentation

### PR DESCRIPTION
**Pull request by erictune**
Work in progress, on documenting 3 different ways to run a job with parallel pods.  I am working on a 4th way, which is the so-called "static work assignment with downward API".  That isn't in this PR.    This PR isn't ready for a final review, but I would like feedback at this point from people on whether it makes sense, if there are things I could do better, and how we feel about there being these different ways to use
job.

CCing people who like jobs:

@soltysh @mikedanese @davidopp @bgrant0607

### CODE REVIEW

File: docs/user-guide/jobs.md [hunk 1]
```diff
@@ -226,6 +228,113 @@ sometimes be started twice.
...
+## Job Patterns
+
+The Job object can be used to support reliable parallel execution of Pods.  The Job object is not
+designed to support closely-communicating parallel processes, as commonly found in scientific
+computing.  It does support parallel processing of a set of independent but related *work items*.
+These might be emails to be sent, frames to be rendered, files to be transcoded, ranges of keys in a
+NoSQL database to scan, and so on.
+
+In a complex system, there may be multiple different sets of work items.  Here we are just
+considering one set of work items that the user wants to manage together &emdash; a *batch job*.
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: `&mdash;` is what you want here.

File: docs/user-guide/jobs.md [hunk 2]
```diff
@@ -226,6 +228,113 @@ sometimes be started twice.
...
+
+There are several different patterns for parallel computation, each with strengths and weaknesses.
+The tradeoffs are:
+
+- One Job object for each work item, vs a single Job object for all work items.  The latter is
+  better for large numbers of work items.  The former creates some overhead for the user and for the
+  system to manage large numbers of Job objects.  Also, with the latter, the resource usage of the job
+  (number of concurrently running pods) can be easily adjusted using the `kubectl scale` command.
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: Why scaling, when this is talking about single job object for all work items?

Comment (soltysh) [inline docs/user-guide/jobs.md]: Nvmd, I was thinking one level lower, whereas you're presenting the idea one level higher, how to use jobs.

File: docs/user-guide/jobs.md [hunk 3]
```diff
@@ -226,6 +228,113 @@ sometimes be started twice.
...
+When you specify completions with `.spec.completions`, each Pod created by the Job controller
+has an identical [`spec`](../devel/api-conventions.md#spec-and-status).  This means that
+all pods will have the same command line and the same
+image, the same volumes, and (almost) the same environment variables.  It is up
+to the user to arrange for the pods to do work on different things.  There are
+several different patterns for parallel jobs, each with strengths and weaknesses.
+The Job object supports   For
+example, the pods might all access a shared work queue service to acquire work
+units.
+-->
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: I like this commented-out block. It presents the basic idea of how jobs work, which is obvious to us (the authors), but not necessarily to and end-user. Please leave it here.

File: docs/user-guide/jobs.md [hunk 4]
```diff
@@ -226,6 +228,113 @@ sometimes be started twice.
...
+### Work Queue with Pod Per Item
+
+The *multiple jobs from template* approach creates one job object for each work item, and so may be
+awkward when there are a very large number of work items.
+
+Also, sometimes you may want to limit the number of pods that are created at the same time, to
+conserve resources, and have the job run until items are completed.
+
+The work-queue-based approaches in this and the following section allow a single Job
+object to manage multiple pods running in parallel, and the number of completions to be
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: Unless you're not talking about job's completions, which I assume you don't, I'd rather try to stay away from the word completions. This suggests, you can modify the completions parameter of a job, which we know is not possible.

File: docs/user-guide/jobs.md [hunk 5]
```diff
@@ -226,6 +228,113 @@ sometimes be started twice.
...
+downward API.
+
+This approach creates a pod per work item, but does not require modification of the program in the
+container image to be work-queue aware.
+
+### Single Job starts Controller Pod
+
+Another pattern is for a single Job to create a pod which then creates other pods, acting as a sort
+of custom controller for those pods.  This allows the most flexibility, but may be somewhat
+complicated to get started with and offers less integration with Kubernetes.
```
Comment (soltysh) [inline docs/user-guide/jobs.md]: We should lay out the possible problems, if we want to show this approach. The one I'm currently thinking of are overlapping pods, the biggest problem imho.

**Comment by soltysh** [comment]
I've started looking at this, but haven't finished yet. I'll be back to it when I finish travelling, later today.

File: examples/job/expansions/README.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+other groups of jobs (these are not shown, but there might be other ones).  The second key (`item`)
+distinguishes individual jobs in the group.  Neither of these keys are special to kubernetes --
+you can pick your own label scheme.  Just make sure each job has a different label selector, from
+every other job.
+
+Next, expand the template into multiple files, one for each item to be processed.
+
+```console
+# Expand files into a temporary directory
+$ mkdir ./tmp
```
Comment (soltysh) [inline examples/job/expansions/README.md]: The ls, and create are using `jobs` dir, but here you're using tmp, which is also with sed.

File: examples/job/expansions/README.md [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+```
+
+Here we use the `-l` option to select all jobs that are part of this
+group of jobs.  (There might be other unrelated jobs in the system that we
+do not care to see.)
+
+We can check on the pods as well using the same label selector:
+
+```console
+# Now check on their pods
```
Comment (soltysh) [inline examples/job/expansions/README.md]: The comment above is sufficient. I suggest removing it from the code sample. The same applies to log example.

File: examples/job/expansions/README.md [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        app: jobexample
+        item: %(name)s
+    spec:
+      containers:
+      - name: c
+        image: busybox
+        command: ["sh", "-c", "echo Processing item %(filename)s in language %(language)s && sleep 5"]
+      restartPolicy: Never
+---
+""" % p)
```
Comment (soltysh) [inline examples/job/expansions/README.md]: I think we should provide at least the few necessary steps to actually run this example. So that a user can just hit download and copy&paste the command and voila have the job running in k8s. Currently this sample is ok, but it's quite raw. So either not include it or improve it, imho. I'd personally go the latter way.

File: examples/job/expansions/README.md [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+[Download example](job-expand.py?raw=true)
+
+
+## Alternatives
+
+If you have a large number of job objects, you may find that:
+
+- even using labels, managing so many Job objects is cumbersome.
+- you exceed resource quota when creating all the Jobs at once,
+  and do not want to wait to create them incrementally
```
Comment (soltysh) [inline examples/job/expansions/README.md]: Missing dots at the end of stmt.

File: examples/job/work-queue-1/README.md [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+1. **Start a message queue service.**  In this example, we use RabbitMQ, but you could use another
+  one.  In practice you would set up a message queue service once and reuse it for many jobs.
+1. **Create a queue, and fill it with messages.**  Each message represents one task to be done.  In
+  this example, a message is just an integer that we will do a lengthy computation on.
+1. **Start a Job that works on tasks from the queue**.  The Job starts several pods.  Each pod takes
+  one task from the message queue, processes it, and repeats until the end of the queue is reached.
+
+## Starting a message queue service
+
+In this example, we use RabbitMQ as a message service.  This example should be easy to adapt to
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: I'd rephrase it like this:
Although this example uses RabbitMQ it should be easy to adapt to another AMQP-type message service.

Currently I'm feeling like reading the same thing about using RabbitMQ twice.

File: examples/job/work-queue-1/README.md [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+another AMQP-type message service.  In practice you could set up a message queue service once in a
+cluster and reuse it for many jobs, as well as for long-running services.
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: The same applies to the last sentence, maybe entirely getting rid of it would be fine?

File: examples/job/expansions/README.md [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+jobexample-banana-bv4s8   0/1       Completed   0          1m
+jobexample-cherry-ml2rv   0/1       Completed   0          1m
+```
+
+There is not a single command to check on the output of all jobs at once,
+but looping over all the pods is pretty easy:
+
+```console
+# Get logs of all the pods.
+$ for p in $(kubectl get pods -l app=jobexample -o name | cut -f 2 -d "/")
```
Comment (soltysh) [inline examples/job/expansions/README.md]: You don't need to cut the pod from the name, log command accepts pods/pod_name.

File: examples/job/expansions/job-expand.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
+#! /usr/bin/python
```
Comment (soltysh) [inline examples/job/expansions/job-expand.py]: First of all you're mixing tabs and spaces in this file, which is highly undesirable in python world. Please use either.

File: examples/job/expansions/job-expand.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+			"args": r"./香蕉.txt",
+		},
+		{
+			"name": "cherry",
+			"args": r"../вишня.doc",
+		}
+	]
+
+for p in params:
+  print (
```
Comment (soltysh) [inline examples/job/expansions/job-expand.py]: I've did slight change to this print line and transformed it into:
```
+    with open("jobs/job-%s.yaml"%p["name"], "w") as f:
+        f.write("""
```
This will create files similarly to what the sed from previous example.

Comment (soltysh) [inline examples/job/expansions/job-expand.py]: With this changes you can write down following usage:
```
mkdir ./jobs
python job-expand.py
kubectl create -f ./jobs
```
and user should be clear what to do from there on.

File: examples/job/work-queue-1/README.md [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+Start RabbitMQ as follows:
+
+```console
+$ kubectl create -f examples/celery-rabbitmq/rabbitmq-service.yaml
+service "rabbitmq-service" created
+$ kubectl create -f examples/celery-rabbitmq/rabbitmq-controller.yaml
+replicationController "rabbitmq-controller" created
+```
+
+We will not use Celery in this example.
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: Positive sounds better, imho:
We will only use the rabbitmq part from the celery-rabbitmq example.

File: examples/job/work-queue-1/README.md [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+do
+  /usr/bin/amqp-publish --url=$BROKER_URL -r job1 -p -b $f
+done
+```
+
+So, we filled the queue with 8 messages.
+
+## Create an Image
+
+Now we are ready to create an image that we will run.
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: we will run as a job.

File: examples/job/work-queue-1/README.md [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+#! /usr/bin/python
+import sys
+print("Processing " + sys.stdin.lines())
+EOF
+$ chmod +x worker.py
+$ cat > Dockerfile <<EOF
+# Specify BROKER_URL and QUEUE when running
+FROM ubuntu:14.04
+
+RUN apt-get update && apt-get install -y curl ca-certificates amqp-tools python
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: This ends up being two separate lines in the actual file, add `\` at the end of this line.

File: examples/job/work-queue-1/README.md [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+--no-install-recommends && rm -rf /var/lib/apt/lists/*
+COPY ./worker.py /worker.py
+
+CMD  /usr/bin/amqp-consume --url=$BROKER_URL -q $QUEUE -c 1 /worker.py
+EOF
+$ cd ..
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: You don't need to go up, just modify build to point to current directory, like so:
```
docker build -t job-wq-1 .
```

File: examples/job/work-queue-1/README.md [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+      restartPolicy: Never
+```
+
+In this example, each pod works on one item from the queue and then exits.
+So, the completion count of the Job corresponds to the number of work items
+done.  So we set, `.spec.completions: 8` for the example, since we put 8 items in the queue.
+
+We set the `.spec.template.spec.restartPolicy` to `Never`.  You could also set it to `OnFailure`.
+
+You'll need to make a copy of the Job and edit the image to match the name you used, and call it
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: I'd move this information before the file contents.

File: examples/job/work-queue-1/README.md [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+      - name: c
+        image: gcr.io/<project>/job-wq-1
+      restartPolicy: Never
+```
+
+In this example, each pod works on one item from the queue and then exits.
+So, the completion count of the Job corresponds to the number of work items
+done.  So we set, `.spec.completions: 8` for the example, since we put 8 items in the queue.
+
+We set the `.spec.template.spec.restartPolicy` to `Never`.  You could also set it to `OnFailure`.
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: Why? Explain in more detail or at least link to where restart policy is discussed.

File: examples/job/work-queue-1/README.md [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+If the number of completions is set to more than the number of items in the queue,
+then the Job will not appear to be completed, even though all items in the queue
+have been processed.  It will start additional pods which will block waiting
+for a mesage.
+
+There is an unlikely race with this pattern.  If the container is killed in between the time
+that the message is acknowledged by the amqp-consume command and the time that the container
+exits with success, or if the node crashes before the kubelet is able to post the success of the pod
+back to the api-server, then the Job will not appear to be complete, even though all items
+in the queue have been processed.
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: I'd add that possible solution to this might be ACK-ing after processing the item instead of before.

File: examples/job/work-queue-2/README.md [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Example: Job with Work Queue with Pod Per Work Item
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: Multiple Items per Pod

Comment (soltysh) [inline examples/job/work-queue-2/README.md]: Nvmd, I thought about the next example.

File: examples/job/work-queue-2/README.md [hunk 23]
```diff
@@ -0,0 +1,282 @@
...
+# Make a Dockerfile
+$ cat > Dockerfile <<EOF
+FROM python
+RUN pip install redis
+COPY ./worker.py /worker.py
+COPY ./rediswq.py /rediswq.py
+
+CMD  python worker.py
+EOF
+$ cd ..
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: No need for cd, you can build from current dir.

File: examples/job/work-queue-2/README.md [hunk 24]
```diff
@@ -0,0 +1,282 @@
...
+Worker with sessionID: bbd72d0a-9e5c-4dd6-abf6-416cc267991f
+Inital queue state: empty=False
+Working on banana
+Working on date
+Working on lemon
+```
+
+As you can see, one of our pods worked on several work units.
+
+**TODO:** extend job controller to wait for excess pods to complete, as created via above.
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: I haven't tried this example yet, I have bad internet connection here... :/

Comment (soltysh) [inline examples/job/work-queue-2/README.md]: Can you elaborate what do you mean with this?

File: examples/job/work-queue-2/README.md [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+**TODO(erictune)**: update above example text once issue fixed with Job controller to allow more pods than completions.
+**TODO**: how does a user or a autoscaler know whether completions is max parallelism
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: User can get that from kubectl get/describe

**Comment by soltysh** [comment]
@erictune good job so far. I really like what you have. I'll be checking your samples once again when I'll have some more reasonable internet connection to do docker push :wink:

File: examples/job/work-queue-1/README.md (hunk 11 above)
Comment (erictune) [inline examples/job/work-queue-1/README.md]: I think this is saying something different from the previous two sentences.

Comment (soltysh) [inline examples/job/work-queue-1/README.md]: By that, I meant it's repeated, it appears first in the overview no 1.

File: examples/job/expansions/README.md (hunk 12 above)
Comment (soltysh) [inline examples/job/expansions/README.md]: I'll check it once again...

File: examples/job/work-queue-1/README.md (hunk 21 above)
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: Right, makes sense. My bad.

File: examples/job/work-queue-2/README.md (hunk 24 above)
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: Why would we want that kind of option? Can you provide a use-case for that?

File: docs/user-guide/jobs.md (hunk 2 above)
Comment (bgrant0607) [inline docs/user-guide/jobs.md]: For static work distribution, I don't know that scaling the number of completions makes sense.

File: docs/user-guide/jobs.md (hunk 3 above)
Comment (bgrant0607) [inline docs/user-guide/jobs.md]: Why is it commented out?

File: docs/user-guide/jobs.md [hunk 26]
```diff
@@ -179,30 +183,28 @@ Also you should not normally create any pods whose labels match this selector, e
...
+running concurrently may be lower or higher for a variety of reasons.  For
+example, it may be lower if the number of remaining completions is less, or as
+the controller is ramping up, or if it is throttling the job due to excessive
+failures.  It may be higher for example if a pod is gracefully shutdown, and
+the replacement starts early.
 
 If you do not specify `.spec.parallelism`, then it defaults to `.spec.completions`.
 
+Depending on the pattern you are using, you will either set `.spec.completions`
+to 1 or to the number of units of work (see [Job Patterns] for an explanation).
```
Comment (bgrant0607) [inline docs/user-guide/jobs.md]: Does this doc explain why I'd use a Job to run one Pod?

File: docs/user-guide/jobs.md [hunk 27]
```diff
@@ -226,6 +228,113 @@ sometimes be started twice.
...
+to the user to arrange for the pods to do work on different things.  There are
+several different patterns for parallel jobs, each with strengths and weaknesses.
+The Job object supports   For
+example, the pods might all access a shared work queue service to acquire work
+units.
+-->
+
+### Work Queue with Pod Per Item
+
+The *multiple jobs from template* approach creates one job object for each work item, and so may be
```
Comment (bgrant0607) [inline docs/user-guide/jobs.md]: Multiple jobs or multiple pods?

File: docs/user-guide/jobs.md [hunk 28]
```diff
@@ -226,6 +228,113 @@ sometimes be started twice.
...
+awkward when there are a very large number of work items.
+
+Also, sometimes you may want to limit the number of pods that are created at the same time, to
```
Comment (bgrant0607) [inline docs/user-guide/jobs.md]: One reason for limiting parallelism is to limit requests to a centralized resource, such as a database for file server.

File: docs/user-guide/jobs.md (hunk 4 above)
Comment (bgrant0607) [inline docs/user-guide/jobs.md]: I suggest "work items" for the more abstract notion.

File: docs/user-guide/jobs.md [hunk 29]
```diff
@@ -226,6 +228,113 @@ sometimes be started twice.
...
+
+Another pattern is for a single Job to create a pod which then creates other pods, acting as a sort
+of custom controller for those pods.  This allows the most flexibility, but may be somewhat
+complicated to get started with and offers less integration with Kubernetes.
+
+One example of this pattern would be a Job which starts a Pod which runs a script that in turn
+starts a Spark master controller (see [spark example](../../examples/spark/README.md)), runs a spark
+driver, and then cleans up.
+
+An advantage of this approach is that the overall process gets the completion guarantee of a Job
```
Comment (bgrant0607) [inline docs/user-guide/jobs.md]: This would be much cleaner with existence dependencies (discussed in #1535, #3585, and elsewhere), since deletion of the Job could then automatically delete the RC and its pods.

File: examples/job/expansions/README.md [hunk 30]
```diff
@@ -0,0 +1,237 @@
...
+      - name: c
+        image: busybox
+        command: ["sh", "-c", "echo Processing item $ITEM && sleep 5"]
+      restartPolicy: Never
+```
+
+[Download example](job-template.yaml?raw=true)
+<!-- END MUNGE: EXAMPLE job-template.yaml -->
+
+Unlike a *pod template*, a job template is not a Kubernetes API type.  It is just
```
Comment (bgrant0607) [inline examples/job/expansions/README.md]: Parameterized templates are being discussed:
https://groups.google.com/d/msg/kubernetes-sig-config/MIyDpX60kUw/OrCwmJPSBwAJ
#10408

File: examples/job/expansions/README.md [hunk 31]
```diff
@@ -0,0 +1,237 @@
...
+sleep for a bit.
+
+In a real use case, the processing would be some substantial computation, such as
+rendering a frame of a movie, or processing a range of rows in a database.
+The "$ITEM" parameter would specify for example, the frame number or the row range.
+
+This Job has two labels. The first label, with key `app`, distinguishes this group of jobs from
+other groups of jobs (these are not shown, but there might be other ones).  The second key (`item`)
+distinguishes individual jobs in the group.  Neither of these keys are special to kubernetes --
+you can pick your own label scheme.  Just make sure each job has a different label selector, from
```
Comment (bgrant0607) [inline examples/job/expansions/README.md]: I'm in favor of generating a unique label by default, as previously discussed. #12298

File: examples/job/expansions/README.md [hunk 32]
```diff
@@ -0,0 +1,237 @@
...
+$ kubectl create -f ./jobs
+job "jobexample-apple" created
+job "jobexample-banana" created
+job "jobexample-cherry" created
+```
+
+Now, check on the jobs:
+
+```console
+$ kubectl get jobs -l app=jobexample
```
Comment (bgrant0607) [inline examples/job/expansions/README.md]: Could also do -Litem

File: docs/user-guide/jobs.md (hunk 4 above)
Comment (soltysh) [inline docs/user-guide/jobs.md]: :+1:

File: examples/examples_test.go [hunk 33]
```diff
@@ -389,6 +389,15 @@ func TestExampleObjectSchemas(t *testing.T) {
...
 		},
+		"../examples/job/expansions": {
+			//"job":   &api.Pod{},
+		},
+		"../examples/job/work-queue-1": {
+			//"job":   &api.Pod{},
+		},
+		"../examples/job/work-queue-2": {
+			//"job":   &api.Pod{},
+		},
```
Comment (soltysh) [inline examples/examples_test.go]: This should be:
```
		"../examples/job/expansions": {
			"job-template": nil,
		},
		"../examples/job/work-queue-2": {
			"redis-pod":     &api.Pod{},
			"redis-service": &api.Service{},
		},
```

Comment (soltysh) [inline examples/examples_test.go]: Unfortunately the job-template needs to be skipped since it's a template and it fails validation. This in turn leads to not checking its syntax, the same applies to job-expand.py.

File: examples/job/expansions/README.md (hunk 12 above)
Comment (soltysh) [inline examples/job/expansions/README.md]: Double checked, the following:
```
for p in $(kubectl get pods -l app=jobexample -o name)
do
  kubectl logs $p
done
```
works without any problems while testing on your branch.

File: examples/job/work-queue-1/README.md [hunk 34]
```diff
@@ -0,0 +1,339 @@
...
+Here is the job definition:
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Job
+metadata:
+  name: job-wq-1
+spec:
+  selector:
+    app: job-wq-1
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: The selector needs to be:
```
  selector:
    matchLabels:
      app: job-wq-1
```

File: examples/job/work-queue-1/README.md [hunk 35]
```diff
@@ -0,0 +1,339 @@
...
+# Note the rabitmq-service has a DNS name, provided by Kubernetes:
+
+root@temp-loe07:/# nslookup rabbitmq-service
+erver:		10.0.0.10
+Address:	10.0.0.10#53
+
+Name:	rabbitmq-service.default.svc.cluster.local
+Address: 10.0.147.152
+
+# Your address will vary.
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: For some reason (?) this step did not work for me. I fall back to getting that ip address from the env variable assigned to the service. Maybe it'll be worth adding this fallback step, or even better the BROKER_URL could use it.

File: examples/job/work-queue-2/README.md [hunk 36]
```diff
@@ -0,0 +1,282 @@
...
+```console
+$ kubectl run -i --tty temp --image redis --command "/bin/sh"
+Waiting for pod default/redis2-c7h78 to be running, status is Pending, pod ready: false
+Hit enter for command prompt
+```
+
+Now hit enter, start the redis CLI, and create a list with some work items in it.
+
+```
+# redis-cli -h redis
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: The domain name again did not work for me.

File: examples/job/work-queue-2/README.md [hunk 37]
```diff
@@ -0,0 +1,282 @@
...
+while not q.empty():
+  item = q.lease()  # Put a timeout here...
+  itemstr = item.decode("utf=8")
+  print("Working on " + itemstr)
+  time.sleep 10 # Put your actual work here instead of sleep.
+  q.complete(item)
+EOF
+$ chmod +x worker.py
+# Add a library for a redis-basd workqueue.
+$ cp examples/job/work-queue-2/rediswq.py ./rediswq.py
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: You're in `tmp/` subdir, so this copy should be:
```
cp ../examples/job/work-queue-2/rediswq.py ./rediswq.py
```

File: examples/job/work-queue-2/README.md [hunk 38]
```diff
@@ -0,0 +1,282 @@
...
+$ cat > Dockerfile <<EOF
+FROM python
+RUN pip install redis
+COPY ./worker.py /worker.py
+COPY ./rediswq.py /rediswq.py
+
+CMD  python worker.py
+EOF
+$ cd ..
+$ docker build -t job-wq-2 ./tmp
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: Again no need to go up, just `docker build -t job-wq-2 . `

File: examples/job/work-queue-2/README.md [hunk 39]
```diff
@@ -0,0 +1,282 @@
...
+Here is the job definition:
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Job
+metadata:
+  name: job-wq-2
+spec:
+  selector:
+    app: job-wq-2
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: The selector needs to be:
```
  selector:
    matchLabels:
      app: job-wq-2
```

File: examples/job/work-queue-2/README.md [hunk 40]
```diff
@@ -0,0 +1,282 @@
...
+import rediswq
+
+q = rediswq.RedisWQ(name="job2", host="redis")
+print("Worker with sessionID: " +  q.sessionID())
+print("Inital queue state: empty=" + str(q.empty()))
+while not q.empty():
+  item = q.lease()  # Put a timeout here...
+  itemstr = item.decode("utf=8")
+  print("Working on " + itemstr)
+  time.sleep 10 # Put your actual work here instead of sleep.
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: You forgot parentheses: `time.sleep(10)`

File: examples/job/work-queue-2/README.md [hunk 41]
```diff
@@ -0,0 +1,282 @@
...
+
+```console
+$ mkdir tmp
+$ cd tmp
+$ cat > worker.py <<EOF
+#! /usr/bin/python
+import time
+import rediswq
+
+q = rediswq.RedisWQ(name="job2", host="redis")
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: Since the hostname resolution did not work for me I had to use env var here again `os.getenv("REDIS_SERVICE_HOST")`

File: examples/job/expansions/README.md (hunk 12 above)
Comment (soltysh) [inline examples/job/expansions/README.md]: Even more you're using that pattern in work-queue-2, towards the end of README.md.

**Comment by soltysh** [comment]
Now I'm done reviewing :)

File: examples/job/work-queue-2/README.md (hunk 24 above)
Comment (bgrant0607) [inline examples/job/work-queue-2/README.md]: That seems pretty confusing.

File: docs/user-guide/jobs.md (hunk 5 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: can you elaborate on what you meant here?

File: docs/user-guide/jobs.md (hunk 26 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: yes

File: docs/user-guide/jobs.md (hunk 27 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: yes.

File: docs/user-guide/jobs.md (hunk 28 above)
Comment (erictune) [inline docs/user-guide/jobs.md]: added

**Comment by erictune** [comment]
Now depends on #18312

**Comment by erictune** [comment]
@soltysh I have addressed all comments.  I think it is ready to go.

File: docs/user-guide/jobs.md (hunk 5 above)
Comment (soltysh) [inline docs/user-guide/jobs.md]: I've meant the usual problems one have to face when implementing controller in k8s. This is sort of controller, although running differently, but the overlapping pods is one example one should be aware of.

File: docs/user-guide/jobs.md (hunk 2 above)
Comment (soltysh) [inline docs/user-guide/jobs.md]: @bgrant0607 for jobs calling changes the `.spec.parallelism` parameter. Completions cannot be changed once submitted.

File: docs/user-guide/jobs.md (hunk 29 above)
Comment (soltysh) [inline docs/user-guide/jobs.md]: @sdminonne those issues might be useful for your workflow proposal, at least sort of.

File: examples/job/expansions/README.md [hunk 42]
```diff
@@ -0,0 +1,255 @@
...
+process-item-apple-kixwv    0/1       Completed   0          4m        apple
+process-item-banana-wrsf7   0/1       Completed   0          4m        banana
+process-item-cherry-dnfu9   0/1       Completed   0          4m        cherry
+```
+
+There is not a single command to check on the output of all jobs at once,
+but looping over all the pods is pretty easy:
+
+```console
+$ for p in $(kubectl get pods -l app=jobexample -o name | cut -f 2 -d "/")
```
Comment (soltysh) [inline examples/job/expansions/README.md]: I'm still disappointed that the version without the cut didn't work for you. I've just checked once again - works perfectly ok.

File: examples/job/expansions/README.md [hunk 43]
```diff
@@ -0,0 +1,255 @@
...
+.)  We can pipe the output directly to kubectl to
+create the objects.
+
+Now, use a one-line python program to expand the template:
+
+```
+$ alias render_template='python -c "from jinja2 import Template; import sys; print(Template(sys.stdin.read()).render());"'
+```
+
+You will need the jinja2 package installed first.
```
Comment (soltysh) [inline examples/job/expansions/README.md]: Maybe add something like this:
`pip install --user jinja2`

Comment (soltysh) [inline examples/job/expansions/README.md]: Though this assumes user has pip installed :/

File: examples/job/work-queue-1/README.md [hunk 44]
```diff
@@ -0,0 +1,359 @@
...
+First create a temporary interactive Pod.
+
+```console
+# Create a temporary interactive container
+$ kubectl run -i --tty temp --image ubuntu:14.04
+Waiting for pod default/temp-loe07 to be running, status is Pending, pod ready: false
+... [ previous line repeats several times .. hit return when it stops ] ...
+```
+
+Note that your podname and command prompt will be different.
```
Comment (soltysh) [inline examples/job/work-queue-1/README.md]: s/podname/pod name

File: examples/job/work-queue-2/README.md [hunk 45]
```diff
@@ -0,0 +1,294 @@
...
+non-parallel, use of [Job](../../../docs/user-guide/jobs.md) first.
+
+In this example, as each pod is created, it picks up one unit of work
+from a task queue, completes it, deletes it from the queue, and exits.
+
+
+Here is an overview of the steps in this example:
+
+1. **Start a storage service to hold the work queue.**  In this example, we use Redis to store
+  our work items.  In the previous example, we used RabbitM.  In this example, we use Redis and
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: s/RabbitM/RabbitMQ

File: examples/job/work-queue-1/Dockerfile [hunk 46]
```diff
@@ -0,0 +1,10 @@
+# Specify BROKER_URL and QUEUE when running
+FROM ubuntu:14.04
+
+RUN apt-get update && \\
```
Comment (soltysh) [inline examples/job/work-queue-1/Dockerfile]: Those double backslashes don't work, you need single backslashes.

File: examples/job/work-queue-2/README.md [hunk 47]
```diff
@@ -0,0 +1,294 @@
...
+
+## Create an Image
+
+Now we are ready to create an image that we will run.
+
+We will use a python worker program with a redis client to read
+the messages from the message queue.
+
+A simple Redis work queue client library is provided,
+called rediswq.py ([Download](rediswq.py)).
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: the link should be `rediseq.py?raw`

File: examples/job/work-queue-2/README.md [hunk 48]
```diff
@@ -0,0 +1,294 @@
...
+    print("Waiting for work")
+print("Queue empty, exiting")
+```
+
+[Download example](worker.py?raw=true)
+<!-- END MUNGE: EXAMPLE worker.py -->
+
+If you are working from the source tree,
+change directory to the `examples/job/work-queue-2` directory.
+Otherwise, download `worker.py`, `redis.py`, and [`Dockerfile`](Dockerfile?format=raw)
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: It might be convenient to have the python files be links as well here.

Comment (soltysh) [inline examples/job/work-queue-2/README.md]: The redis.py is actually rediswq.py.

Comment (soltysh) [inline examples/job/work-queue-2/README.md]: The link should be `Dockerfile?raw=true`.

File: examples/job/work-queue-2/README.md [hunk 49]
```diff
@@ -0,0 +1,294 @@
...
+Working on banana
+Working on date
+Working on lemon
+```
+
+As you can see, one of our pods worked on several work units.
+
+**TODO:** extend job controller to wait for excess pods to complete, as created via above.
+**TODO(erictune)**: update above example text once issue fixed with Job controller to allow more pods than completions.
+**TODO**: kill a pod or even reboot a node while job is running and have the job still complete.
```
Comment (soltysh) [inline examples/job/work-queue-2/README.md]: If you don't mind I'd rather move those todos as issues, and having this file clean. I'm seeing this as an unnecessary stuff for people going through the examples.

**Comment by soltysh** [comment]
Nits here and there, but mostly LGTM. Once you have those fixed (pretty please), you can apply the lgtm label :)

File: examples/job/expansions/README.md (hunk 42 above)
Comment (erictune) [inline examples/job/expansions/README.md]: I rebuilt kubectl and now it works.  changed to your way.

**Comment by erictune** [comment]
Thanks for the careful reviews!  Fixed all your comments.

File: examples/job/expansions/README.md (hunk 42 above)
Comment (soltysh) [inline examples/job/expansions/README.md]: :+1:

**Comment by soltysh** [comment]
Thank you for putting it all together! One last nit that is the boilerplate for `work-queue-1/worker.py`, `work-queue-2/worker.py` and `work-queue-2/rediswq.py` is not correct :wink: and you're good to go!

**Comment by erictune** [comment]
Boilerplate fixed.

**Comment by bgrant0607** [comment]
FYI, these examples belong in the user-guide. /examples is for actual apps.
